### PR TITLE
ROI Color Palette

### PR DESCRIPTION
**Pull request by barrettMCW**
I finally got around to fixing this up. I did a fair bit of testing this time, and as far as I can tell it works as intended. 

EDIT (Will): testing: the following config is now on merge-ci and this PR should be included, so the ROI color-picker should correspond to those colors:

```
omero config set omero.web.iviewer.roi_color_palette "[rgb(0,255,0),rgb(100,50,150)],[darkred,red,pink],[#0000FF, #ff99ff]"
omero config set omero.web.iviewer.show_palette_only true
```

Screenshot 2022-08-02 at 22 31 20

### CODE REVIEW

**Comment by jburel** [comment]
```
flake8.main.application   MainProcess    760 INFO     Found a total of 13 violations and reported 1
/omero-iviewer/plugin/omero_iviewer/iviewer_settings.py:43:1: W293 blank line contains whitespace
```

File: README.rst [hunk 1]
```diff
@@ -73,6 +73,18 @@ If you wish to set a threshold for iviewer that is *lower* than for the server:
 NB: Z-projection is not supported for tiled images in OMERO
 (Images larger than 2048 * 2048 pixels per plane are tiled in iviewer).
 
+OMERO uses Spectrum Color Picker for selecting ROI colors. 
+Define rows with brackets, and use commas to seperate values. 
+You can only view the first item of each row without enabling show_palette_only.
+To define a color palette use::
+    
+    $ omero config set omero.web.iviewer.color_palette [rgb(0,0,0),blue],[#000000]
```
Comment (will-moore) [inline README.rst:81]: The description here doesn't give me much info on what this setting does. After the first statement you need something like, "The `color_palette` setting allows you to specify a grid of colors for users to choose for ROIs".
Typo "separate". The config name doesn't include 'roi' anywhere. I wonder if a better name might be `roi_colors` or `roi_palette` or even `roi_color_palette`?

**Comment by will-moore** [comment]
I tried the suggested setting from the README:

```
$ omero config set omero.web.iviewer.roi_color_palette [rgb(0,0,0),blue],[#000000]
bash: syntax error near unexpected token `('
```

Then with double quotes:

```
$ omero config set omero.web.iviewer.roi_color_palette "[rgb(0,0,0),blue],[#000000]"
```

But this doesn't seem to be parsed correctly (just gives 2 black options + plus the default/current yellow):

Screenshot 2022-07-21 at 14 48 16

I wonder if it would be simpler to use `json` instead of string for the config value. E,g. https://github.com/ome/omero-web/blob/a34be15e664251bfe5decc1aa8f15a7f4801661b/omeroweb/settings.py#L994

Then setting would be like this (and the parsing happens when you start `omero-web`):

```
$ omero config set omero.web.iviewer.roi_color_palette '[["rgb(0,0,0)","blue"],["#000000"]]'
```

**Comment by barrettMCW** [comment]
That would be mostly correct actually. The yellow shouldn't be there, that must have broke when I changed variable names, I'm looking at it now. But with the config you set both the first items are black. The blue isn't displaying because it's the second item, you'd need to show_palette_only to see the blue. I'll fix the yellow and take another shot at the readme

File: README.rst [hunk 2]
```diff
@@ -73,6 +73,18 @@ If you wish to set a threshold for iviewer that is *lower* than for the server:
 NB: Z-projection is not supported for tiled images in OMERO
 (Images larger than 2048 * 2048 pixels per plane are tiled in iviewer).
 
+OMERO uses Spectrum Color Picker for selecting ROI colors. 
+Define rows with brackets, and use commas to seperate values.
```
Comment (will-moore) [inline README.rst:78]: Typo: "seperate" -> "separate"

File: README.rst [hunk 3]
```diff
@@ -73,6 +73,19 @@ If you wish to set a threshold for iviewer that is *lower* than for the server:
 NB: Z-projection is not supported for tiled images in OMERO
 (Images larger than 2048 * 2048 pixels per plane are tiled in iviewer).
 
+OMERO uses Spectrum Color Picker for selecting ROI colors. 
+The roi_color_palette option allows you to specify a grid of colors for users to choose for ROIs.
+Define rows with brackets, and use commas to seperate values. 
+You can only view the first item of each row without enabling show_palette_only.
```
Comment (will-moore) [inline README.rst:79]: Hopefully this is a bit clearer suggestion: "By default, only the first color of each row is shown. A full grid is shown when the default color picker is hidden (see below)".

File: README.rst [hunk 4]
```diff
@@ -73,6 +73,19 @@ If you wish to set a threshold for iviewer that is *lower* than for the server:
...
+To define a color palette use::
+    
+    $ omero config set omero.web.iviewer.roi_color_palette "[rgb(0,255,0)],[darkred,red,pink],[#0000FF]"
+  
+Hides the default color picker.
```
Comment (will-moore) [inline README.rst:84]: "Hides the default color picker." This seems strange when you're referring to the setting below (this isn't a title).
Maybe: "To hide the default color picker (and show a grid for the color palette), set show_palette_only to true".

**Comment by will-moore** [comment]
I tried using `$ omero config set omero.web.iviewer.roi_color_palette "[rgb(0,255,0),rgb(100,50,150)],[darkred,red,pink],[#0000FF, #ff99ff]"` but the first row got treated as a single string `rgb(0,255,0),rgb(100,50,150)`

File: src/model/regions_info.js [hunk 5]
```diff
@@ -728,7 +727,13 @@ export default class RegionsInfo  {
      * @memberof RegionsInfo
      */
     resetShapeDefaults() {
-        this.shape_defaults['StrokeColor'] = -65281;
+        //if they set a color palette, set the first one as default stroke color
+        console.log(this.image_info.context.roi_color_palette)
```
Comment (will-moore) [inline src/model/regions_info.js:731]: Don't need to console.log this all the time (although it was helpful for my debugging). OK to log it IF it's actually set.

**Comment by barrettMCW** [comment]
Apologies, missed the console.log and I'm bad with words. Good catch, I fixed the parsing regex. Thanks for your help

**Comment by will-moore** [comment]
```
omero.web.iviewer.roi_color_palette=[rgb(0,255,0)],[rgb(0,50,150),darkred,rgb(110,150,0),red],[#0000FF, #ff99ff, rgb(100,50,150)]
omero.web.iviewer.show_palette_only=true
```
It's looking good now, thanks:

Screenshot 2022-08-01 at 16 57 41

All my testing has been local so far. I'll give it a final go on the merge-ci server tomorrow...

**Comment by pwalczysko** [comment]
Works fine on merge-ci. Like the intuitiveness of the feature. LGTM

**Review by pwalczysko — APPROVED** [review]

**Comment by jburel** [comment]
As discussed this morning, it will be good to test the code **without** the configuration described above to see if there is any side effects.

**Comment by will-moore** [comment]
@jburel I removed the config on merge-ci (for tomorrow), so if you want to test this PR *with* config, it would be best to do it today.

**Comment by jburel** [comment]
Tested without the setting and the viewer works as expected

**Pull request merged**